src/tss2-tcti/tcti-cmd.c

```diff
@@ -0,0 +1,453 @@
+   	                1, TPM_HEADER_SIZE, tcti_cmd->source);
+   	if (ferror(tcti_cmd->source)) {
+    	LOG_ERROR("Reading from command TCTI: %s", strerror(errno));
+    	rc = TSS2_TCTI_RC_IO_ERROR;
+    	goto out;
+    }
+
+	if (size != TPM_HEADER_SIZE) {
+		LOG_ERROR("Read was not size of header");
+		rc = TSS2_TCTI_RC_IO_ERROR;
+		goto out;
+	}
+
+	/*
+	 * we have a header, get the size field to compute the rest of
+	 * the data.
+	 */
+	rc = header_unmarshal (
+	    response_buffer,
+	    &tcti_common->header);
+	if (rc) {
+		goto out;
+	}
+
+	/*
+	 * Read the remaining data that is past the header size
+	 */
+	size_t data_size = tcti_common->header.size - TPM_HEADER_SIZE;
+
+	size = fread(&response_buffer[TPM_HEADER_SIZE],
+   	                1, data_size, tcti_cmd->source);
+   	if (ferror(tcti_cmd->source)) {
+    	LOG_ERROR("Reading from command TCTI: %s", strerror(errno));
+    	rc = TSS2_TCTI_RC_IO_ERROR;
+    	goto out;
+    }
+
+   	if (size != data_size) {
+		LOG_ERROR("Write was not size of header");
+		rc = TSS2_TCTI_RC_IO_ERROR;
+		goto out;
+   	}
+
+	*response_size = tcti_common->header.size;
+
+    /*
+     * Executing code beyond this point transitions the state machine to
+     * TRANSMIT. Another call to this function will not be possible until
+     * another command is sent to the TPM.
+     */
+out:
+    tcti_common->header.size = 0;
+    tcti_common->state = TCTI_STATE_TRANSMIT;
+
+    return rc;
+}
+
+void
+tcti_cmd_init_context_data (
+    TSS2_TCTI_COMMON_CONTEXT *tcti_common)
+{
+    TSS2_TCTI_MAGIC (tcti_common) = TCTI_COMMAND_MAGIC;
+    TSS2_TCTI_VERSION (tcti_common) = TCTI_VERSION;
+    TSS2_TCTI_TRANSMIT (tcti_common) = tcti_cmd_transmit;
+    TSS2_TCTI_RECEIVE (tcti_common) = tcti_cmd_receive;
+    TSS2_TCTI_FINALIZE (tcti_common) = tcti_cmd_finalize;
+    TSS2_TCTI_CANCEL (tcti_common) = tcti_cmd_cancel;
+    TSS2_TCTI_GET_POLL_HANDLES (tcti_common) = tcti_cmd_get_poll_handles;
+    TSS2_TCTI_SET_LOCALITY (tcti_common) = tcti_cmd_set_locality;
+    TSS2_TCTI_MAKE_STICKY (tcti_common) = tcti_make_sticky_not_implemented;
+    tcti_common->state = TCTI_STATE_TRANSMIT;
+    tcti_common->locality = 0;
+    memset (&tcti_common->header, 0, sizeof (tcti_common->header));
+}
+
+static bool popen_w_pipes(const char *cmd, pid_t *pid,
+		FILE **sink, FILE **source, FILE **err) {
+
+	/* TODO either return errno or preserve it for the caller */
+
+	pid_t _pid = 0;
+	int stdout_pipefd[2];
+	int stderr_pipefd[2];
+	int stdin_pipefd[2];
+
+	pipe(stdin_pipefd);
+    pipe(stdout_pipefd);
+    pipe(stderr_pipefd);
+
+    _pid = fork();
+	if (_pid == 0) {
+	    dup2(stdin_pipefd[0],  STDIN_FILENO);
+	    dup2(stdout_pipefd[1], STDOUT_FILENO);
+	    dup2(stderr_pipefd[1], STDERR_FILENO);
+
+	    pid_t p = getpid();
+	    (void)p;
+
+	    // ask kernel to deliver SIGTERM in case the parent dies
+	    prctl(PR_SET_PDEATHSIG, SIGTERM);
+
+	    //replace tee with your process
+	    execlp("bash", "bash", "-c", cmd, (char*) NULL);
```

Comment:
Is it a good idea to assume bash? Just "sh" might be safer.

Reply:
Yep, I ended up modeling system(3): `execlp("/bin/sh", "sh", "-c", "cmd", NULL);`